src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs

```diff
@@ -98,6 +107,47 @@ public unsafe int InflateVerified(byte* bufPtr, int length)
             }
         }
 
+        /// <summary>
+        /// If this stream has some input leftover that hasn't been processed then we should
+        /// check if it is another GZip file concatenated with this one.
+        /// 
+        /// Returns false if the leftover input is another GZip data stream.
+        /// </summary>
+        private unsafe bool ResetStreamForLeftoverInput()
+        {
+            Debug.Assert(!NeedsInput());
+            Debug.Assert(IsGzipStream());
+            Debug.Assert(_inputBufferHandle.IsAllocated);
+
+            lock (SyncLock)
+            {
+                IntPtr nextInPtr = _zlibStream.NextIn;
+                byte* nextInPointer = (byte*)nextInPtr.ToPointer();
+                uint nextAvailIn = _zlibStream.AvailIn;
+
+                // Check the leftover bytes to see if they start with he gzip header ID bytes
+                if (*nextInPointer != ZLibNative.GZip_Header_ID1 || (nextAvailIn > 1 && *(nextInPointer + 1) != ZLibNative.GZip_Header_ID2))
+                {
+                    return true;
+                }
+
+                // Trash our existing zstream.
+                _zlibStream.Dispose();
+
+                // Create a new zstream
+                InflateInit(_windowBits);
+
+                // SetInput on the new stream to the bits remaining from the last stream
+                _zlibStream.NextIn = nextInPtr;
+                _zlibStream.AvailIn = nextAvailIn;
+                _finished = false;
+            }
+
+            return false;
+        }
+
+        internal bool IsGzipStream() => _windowBits >= 24 && _windowBits <= 31;
```

Comment:
Maybe it's fine, but this seems a bit unreliable... is it possible in the future we might allow a non-default windowsBits to be specified and end up making this logic wrong?

Reply:
The window bits range is defined by zlib, and this range signals the inflater to treat the stream as gzip:

ranges are:
- -8..-15: deflate
- 0: zlib
- 8..15: zlib
- 24..31: gzip
- 40..47: zlib or gzip (detects header)

Reply:
Ok